Change sendMany call to receive generic JSONObject

charlycrs opened the pull request:
Changed the whole method to receive a JSONObject, I replaced it but I kept thinking if maybe could keep both forms, one with explicit params and one with the json and add an internal private sendMany to keep support for both, but I don't know if this is excesive? What you guys think?

kushal256 commented:
Occasionally for testing we also send parameters not listed in documentation.  Additionally, there appears to be a JSONObject::toMap() call which will translate the JSONObject to a Map<String, Object>

Since this implementation within sendMany() is manually converting from JSONObject to a Map<String, Object>, can we instead just call JSONObject::toMap(), which will also add any parameters not listed in the documentation?  This method implementation might become much simpler.

I'm open to whether or not this method takes a JSONObject or a Map<String, Object> as it's parameter - only downside with taking the JSONObject is we add a new external dependency.  

Also could be nice in our tester to exclude a required parameter to see what the error message is, and capture that in a test case.

What do you guys think?

charlycrs commented:
> Occasionally for testing we also send parameters not listed in documentation. Additionally, there appears to be a JSONObject::toMap() call which will translate the JSONObject to a Map<String, Object>
> 
> Since this implementation within sendMany() is manually converting from JSONObject to a Map<String, Object>, can we instead just call JSONObject::toMap(), which will also add any parameters not listed in the documentation? This method implementation might become much simpler.
> 
> I'm open to whether or not this method takes a JSONObject or a Map<String, Object> as it's parameter - only downside with taking the JSONObject is we add a new external dependency.
> 
> Also could be nice in our tester to exclude a required parameter to see what the error message is, and capture that in a test case.
> 
> What do you guys think?

You might be right, to avoid adding dependencies maybe receive a Map and send from the other side the JSON.toMap. We already know needed parameters so map them out it's quite easy.

I will add a wrong test case for when a needed info is lacking

charlycrs commented:
Done, changed the JSON to HashMap and added a small test for 2 failures, one of required info that goes in the URL and one of required info that goes in the payload.  Maybe later would be good idea to add to the test one of a misformed optional or required info. @kushal256 could you take another look to see if I after this I can release and change the other calls from the original way to the HashMap call?

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClient.java:
```diff
@@ -39,26 +38,23 @@
-                                         BigDecimal fee, BigDecimal feeTxConfirmTarget,
-                                         int minConfirms, boolean enforceMinConfirmsForChange)
-            throws IOException;
+    Optional<SendCoinsResponse> sendMany(Map<String, Object> parameters) throws IOException;
```
This is not good design. If there are too many parameters, it's best to create a class with the parameters that can be passed.

Also, if there are required parameters, I think at least those should be passed by name here, and then an object with the optional parameters. Or make the params object validate the required parameters with lombok's non-null annotation.

chochos replied:
If bitgo's API changes too much or is very informal and lets you send additional parameters depending on whatever, then I guess a Map is inevitable, but I do think required parameters (and well-known optional ones) should be explicitly declared somehow.

charlycrs replied:
Yeah, I think you're right. I'm now thinking to return to the previous way and go from there, and maybe having the required ones as they were, one by one and maybe a map for the optional ones, enhance the javadoc to include a small list of the optional parameters and reinforce all the errors before the request, required that are empty or not there at all and misformed ones to return error without making the request.

chochos replied:
For required vs optional parameters, the root problem is a language limitation (Java doesn't have defaulted parameters). You're not the first person to try and work your way around it; maybe take a look at what others have done before. IIRC the usual workaround is a parameters object with the default values already set, so you can instantiate it and just set whatever you need then pass it to the method.

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -104,14 +104,28 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+        // Get all needed parameters
+        String coin = parameters.get("coin").toString();
+        String walletId = parameters.get("walletId").toString();
+        String walletPass = parameters.get("walletPass").toString();
```
See, the doc specifies these are required parameters, but I don't see any validation here to enforce that (I know, it's being delegated to bitgo, but why do that if it can be enforced beforehand?) – if a parameter is missing from the map there will be a NPE which sheds no light on what's missing, and if the value is present but empty, we're sending an invalid request to bitgo.

chochos commented on bitgo-java-api/src/test/java/com/bitso/bitgo/util/TestClientV2.java:
```diff
@@ -52,14 +53,51 @@ public void testSendMany() throws IOException {
+        // avoid coin required parameter to test error handling
+        parameters.put("coin", "");
+        parameters.put("walletId", WALLET_ID);
+        parameters.put("walletPass", WALLET_PASSPHRASE);
```
There needs to be a test with a missing parameter (not an _empty_ one like here, but no key for that parameter at all).

charlycrs commented:
Changed the method to receive the main parameters as explicit, added NonNull annotations to those.
Optional parameters are map of string objects. 
Added validations for all parameters.
Changed the test for known failures.

I still have a couple of questions: 

Original method has fee sent as `fee.longValue()`, I did not change it but I don't know if some precision can be lost, the other BigDecimal parameters stay BigDecimal.
There is a log `log.info("sendMany {}", data);` this after data gets the walletPassphrase, I think this should go before or not at all to avoid sending this info to a log.
I checked for coin type, but Bitgo can be used for more than those coins, so I don't know if add all or just check for non empty values.
For addresses I don't know if check if it's a valid address or leave this to Bitgo, as the address patterns are different from coin to coin.

chochos commented:
I _think_ fee must be sent in satoshis, but check the api doc of bitgo

chochos commented:
This is meant to be a way to easily call the bitgo api, so validating addresses or currencies is out of scope at least for now.

charlycrs commented:
> I _think_ fee must be sent in satoshis, but check the api doc of bitgo

Yes you're right In the docs, everything regarding amounts fees and all is represented in satoshis. But for our inner use we make the conversion previously. Perhaps add in the docs that amounts are satoshis. My worry was loss of precision when converting from BigDecimal to Long. I don't think it's possible that a fee is that high enough to break break conversion and return opposite sign.

kushal256 commented:
Does TestClientV2 compile?  Looks like a compilation error here:

https://github.com/bitsoex/bitso-bitgo-java-api/blob/bd5181d41b4b94dd97f9ffba81bbb551aaefd7cb/bitgo-java-api/src/test/java/com/bitso/bitgo/util/TestClientV2.java#L116

kushal256 commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -121,25 +135,56 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+                    optionalParameters.get("feeTxConfirmTarget") : null;
+            Object minConfirms = optionalParameters.containsKey("minConfirms") ? optionalParameters.get("minConfirms") : null;
+            Object enforceMinConfirmsForChange = optionalParameters.containsKey("enforceMinConfirmsForChange") ?
+                    optionalParameters.get("enforceMinConfirmsForChange") : null;
```
We also need to support optionalParameters not in their documentation.

Can we add all the optionalParameters to the `data` hashmap?

charlycrs replied:
Ok, sure, just question: Add the whole optionalParameters map after validating the ones we know instead of adding them one by one as I did? Or add the ones known like this and add the map in case there are remaining unknown parameters to us?

charlycrs replied:
> Does TestClientV2 compile? Looks like a compilation error here:
> 
> https://github.com/bitsoex/bitso-bitgo-java-api/blob/bd5181d41b4b94dd97f9ffba81bbb551aaefd7cb/bitgo-java-api/src/test/java/com/bitso/bitgo/util/TestClientV2.java#L116

Really? Crap, I just did a build and everything was alright. Which error did you get?

kushal256 replied:
> Ok, sure, just question: Add the whole optionalParameters map after validating the ones we know instead of adding them one by one as I did? Or add the ones known like this and add the map in case there are remaining unknown parameters to us?

These two options sound the same to me?  If we validate the optional parameters that we know, and add any unknown ones, then i think we're good.

kushal256 replied:
> > Does TestClientV2 compile? Looks like a compilation error here:
> > https://github.com/bitsoex/bitso-bitgo-java-api/blob/bd5181d41b4b94dd97f9ffba81bbb551aaefd7cb/bitgo-java-api/src/test/java/com/bitso/bitgo/util/TestClientV2.java#L116
> 
> Really? Crap, I just did a build and everything was alright. Which error did you get?

Note it's the test, not the src code, that isn't compiling for me.  That line I referenced fails b/c it cannot `Cannot resolve method 'isEmpty' on 'Optional'`

kushal256 replied:
Ahh nevermind, that error was my fault - my project was on jdk8.  When I move to jdk11 it's ok

charlycrs replied:
> > Ok, sure, just question: Add the whole optionalParameters map after validating the ones we know instead of adding them one by one as I did? Or add the ones known like this and add the map in case there are remaining unknown parameters to us?
> 
> These two options sound the same to me? If we validate the optional parameters that we know, and add any unknown ones, then i think we're good.

haha no, my bad, my question was wrong, I meant if validate and then send to Bitgo a HashMap either with all the optionals or just with the unknowns, but as I check this would kill bitgo as it won't know what to do with a map of parameters. So I'll leave the validation but also loop through all the keys and add them to the data. And I guess it's on the user if a bad parameter is sent.

kushal256 commented:
> > I _think_ fee must be sent in satoshis, but check the api doc of bitgo
> 
> Yes you're right In the docs, everything regarding amounts fees and all is represented in satoshis. But for our inner use we make the conversion previously. Perhaps add in the docs that amounts are satoshis. My worry was loss of precision when converting from BigDecimal to Long. I don't think it's possible that a fee is that high enough to break break conversion and return opposite sign.

Why don't we just do BigDecimal::toBigInteger::toString()?  Then we never have to worry about going over Long.MAX_VALUE.

charlycrs commented:
Done, changed the validation of optionals to sweep all of them and validate the ones we know. Also changed the Long to BigInteger just to be sure. @kushal256 could you take another look?

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -104,42 +105,96 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+                                                @NonNull Map<String, BigDecimal> recipients, @NonNull String sequenceId,
+                                                Map<String, Object> optionalParameters) throws IOException {
+        // Validate all needed parameters
+        if (coin.equals("")) {
```
or `coin.isBlank()`

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -104,42 +105,96 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+                optionalValue = currentParameter.getValue();
+                if (optionalKey.equals("message")) {
+                    if (optionalValue != null) {
+                        if (optionalValue.getClass().getName().equals(String.class.getName())) {
```
Why not just call `toString()` on the value?

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -104,42 +105,96 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+                    }
+                } else if (optionalKey.equals("fee")) {
+                    if (optionalValue != null) {
+                        if (optionalValue.getClass().getName().equals(BigDecimal.class.getName())) {
```
`optionalValue instanceof BigDecimal`

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -104,42 +105,96 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+                    }
+                } else if (optionalKey.equals("feeTxConfirmTarget")) {
+                    if (optionalValue != null) {
+                        if (optionalValue.getClass().getName().equals(BigDecimal.class.getName())) {
```
instanceof

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -104,42 +105,96 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+                    }
+                } else if (optionalKey.equals("minConfirms")) {
+                    if (optionalValue != null) {
+                        if (optionalValue.getClass().getName().equals(Integer.class.getName()) && ((int) optionalValue > 0)) {
```
instanceof

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -104,42 +105,96 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+                    }
+                } else if (optionalKey.equals("enforceMinConfirmsForChange")) {
+                    if (optionalValue != null) {
+                        if (optionalValue.getClass().getName().equals(Boolean.class.getName())) {
```
instanceof

chochos commented on bitgo-java-api/src/main/java/com/bitso/bitgo/v2/BitGoClientImpl.java:
```diff
@@ -104,42 +105,96 @@ public BitGoClientImpl(String token, String baseUrl, boolean unsafe) {
+                    }
+                } else {
+                    // Unknown parameter for validation, if not null allow it to pass to the transaction request
+                    if (optionalValue != null) {
```
this can simply be `else if (optionalValue != null) {`
